Improve tests in `python_examples.py`.

Vector73 opened the pull request:
<!-- Describe your pull request here.-->
Adds `assert` statements to check if the request succeeded before
validating against the openapi schema.

Improves the robustness of tests in `python_examples.py`.
Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/openapi/python_examples.py:
```diff
@@ -309,48 +309,54 @@ def get_invitations(client: Client) -> None:
         "invite_expires_in_minutes": 60 * 24 * 10,  # 10 days
         "invite_as": 400,
-        "stream_ids": [1, 11, 12],
+        "stream_ids": stream_ids,  # [10, 12, 3]
```
So I just want to emphasize that the block inside the `{code_example|start}` tags is what appears in the user-facing documentation. So generally we're going to want to put any preparatory requests outside those blocks, and try to make those examples look as clean/clear as possible.

timabbott replied:
I don't think we actually care what the IDs are, so I don't think the comment is required?

Vector73 replied:
I actually added that comment to emphasize that `stream_ids` is a list. Should I remove it?

Vector73 replied:
OK, I have removed that comment.

gnprice commented on zerver/openapi/python_examples.py:
```diff
@@ -154,26 +155,38 @@ def get_presence(client: Client) -> None:
     # Add a stream to the set of default streams for new users.
-    stream_id = 10
-
     result = client.add_default_stream(stream_id)
     # {code_example|end}
 
+    assert result["result"] == "success"
```
This is a good set of changes given that these asserts are needed.

But it also seems like a pretty bad library design choice that these asserts would be needed. What I think a user would expect when using an API binding library like this, in Python or most other languages, is that if the request failed — if it came back with a 4xx or 5xx response instead of 2xx — then the call should throw, not return a result. Relying on the caller to check for an error seems highly error-prone.

(That doesn't mean we should necessarily invest resources in fixing this anytime soon. I'm not sure how many people use these Python bindings.)

laurynmm replied:
It looks like[ 5xx responses do throw](https://github.com/zulip/python-zulip-api/blob/20ccb221193abc283295ce0f69fc78cd2267e800/zulip/zulip/__init__.py#L639-L677), but 4xx responses do not.

There are some 4xx errors checked here and notably `validate_against_openapi_schema` will always return `True` without checking the spec for those cases ...

[CODE LINK](https://github.com/zulip/zulip/blob/1728a2c057166d616fbe677d74e792dc19d65c99/zerver/openapi/openapi.py#L468-L473)
```python
    if status_code.startswith("4"):
        # This return statement should ideally be not here. But since
        # we have not defined 400 responses for various paths this has
        # been added as all 400 have the same schema.  When all 400
        # response have been defined this should be removed.
        return True
```

So, in addition to adding these asserts for "success", it might be good to add a commit to assert for "error" in those 4xx tests. Or add a helper function here that checks/asserts the "result" field in the response, which defaults to checking for "success" and checks for "error" if we're explicitly testing an error response.

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -154,26 +155,38 @@ def get_presence(client: Client) -> None:
+            },
+        ],
+    )
+    stream_id = client.get_stream_id("test stream")["stream_id"]
```
Let's not use "stream" here since we're in the process of renaming stream to channel.

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -309,48 +330,56 @@ def get_invitations(client: Client) -> None:
 @openapi_test_function("/invites:post")
 def send_invitations(client: Client) -> None:
+    streams = client.get_subscriptions()
+    stream_ids = [stream["stream_id"] for stream in streams["subscriptions"]][:3]
```
This is repeated a number of times, so maybe a helper function would be useful?

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -468,18 +504,26 @@ def get_realm_linkifiers(client: Client) -> None:
     # Reorder the linkifiers in the user's organization.
-    order = [4, 3, 2, 1]
+    order = reordered_linkifiers
     request = {"ordered_linkifier_ids": json.dumps(order)}
```
We could probably just pass `reordered_linkifiers` to `json.dumps` here.

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -492,18 +536,27 @@ def get_realm_profile_fields(client: Client) -> None:
     # {code_example|start}
     # Reorder the custom profile fields in the user's organization.
-    order = [9, 8, 7, 6, 5, 4, 3, 2, 1]
+    order = reordered_profile_fields
```
Same note as above about just passing the variable instead of renaming it.

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -534,14 +588,15 @@ def add_realm_filter(client: Client) -> None:
     # {code_example|start}
-    # Update the linkifier (realm_filter) with ID 4
-    filter_id = 4
+    # Update the linkifier (realm_filter)
```
Let's update this to have punctuation and add a period at the end of this sentence. And do the same with any other of these comments updated in the PR.

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -896,54 +969,65 @@ def mark_all_as_read(client: Client) -> None:
     # {code_example|start}
     # Update the user's subscription of the channel with ID 1 so that
     # it's pinned to the top of the user's channel list, and in the
     # channel with ID 3 so that it has the hex color "f00".
```
This comment needs to be updated for your changes.

laurynmm reviewed (commented):
@Vector73 - Thanks for making these improvements. I had a few thoughts. Let me know if you have any questions about my comments.

Also, see my separate comment related to Greg's note. I think it'd be good to check both the success and error tests, so maybe instead of adding all of these asserts, we add a helper function to do that check.

Vector73 commented:
I have addressed the feedbacks by @laurynmm. I have also fixed some incorrect `validate_against_openapi_schema(..., 400)` which I found while verifying the "error" response.

Also, there are 2 functions - `test_user_account_deactivated` and `test_realm_deactivated` - which don't seem to work as expected because `result["result]="success"` but the request is supposed to fail (`REALM_DEACTIVATED` or `USER_DEACTIVATED` error) with 401 status code (the current code though checks for 403 status code which has been changed to 401 for these errors).

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -430,28 +476,31 @@ def update_user(client: Client) -> None:
         url=f"/users/{user_id}/subscriptions/{stream_id}",
         method="GET",
     )
+    validate_response_result(result)
```
This is in the user-facing documentation...

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -766,11 +843,12 @@ def remove_subscriptions(client: Client) -> None:
-    assert result["result"] == "success"
+    validate_response_result(result)
     streams = [s for s in result["subscriptions"] if s["name"] == "python-test"]
     assert len(streams) == 0
```
I'd change this comment to be "Confirm the user is no longer subscribed to "python-test."

Since we're not validating the endpoint call here ... I wonder if we need the result validation? Maybe revise to ...
```python
    # Confirm user is no longer subscribed to "python-test".
    subscriptions = client.get_subscriptions()["subscriptions"]
    streams = [s for s in subscriptions if s["name"] == "python-test"]
    assert len(streams) == 0
```

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -1183,7 +1293,7 @@ def update_message(client: Client, message_id: int) -> None:
     )
-    assert result["result"] == "success"
+    validate_response_result(result)
     assert result["raw_content"] == request["content"]
```
There are three places that we call `GET messages/{message_id}` to compare the `raw_content` of the message with the request message `content`. Seems like another place we could add a helper function.

Vector73 replied:
Created the `validate_message` helper function.

laurynmm reviewed (commented):
@Vector73 - Here's another round of comments for these changes.

I'll note that this would be easier to review if you possible split this into more than one commit. And doing that would also possibly help you self-review for some of the small inconsistencies I've noticed in this last round of review.

So, having a commit for updating the linkifier endpoints to pass the `filter_id` parameter for example. And a separate commit for the helper function for getting subscription IDs. And so on.

Looking at how I split things out in #30135 might be helpful with that.

---

For those two error checks that are returning success:
- I noted that if I comment out all the other tests in `tools/api-tests` and just run the deactivate realm test, then I do get the error response ... though the error code isn't the one documented and as you noted it's a 401 error now :/
- I feel like you can work on figuring out those in a separate follow-up PR since they're already not working

---

For error checks here in general, I'm wondering if we:
- Remove the calls to `validate_against_openapi_schema` since that's not currently validating the error response anyway.
- Add a check for the expected code in the error result.

I'll start a discussion in CZO and/or file an issue for validating error responses, but unfortunately it's a bit more complicated than adding error documentation to all the documented endpoints.

laurynmm commented on zerver/openapi/python_examples.py:
```diff
@@ -430,28 +476,31 @@ def update_user(client: Client) -> None:
     # {code_example|start}
     # Check whether a user is a subscriber to a given channel.
     user_id = 7
-    stream_id = 1
```
Probably can be done as a follow-up PR to these changes.

Vector73 commented on zerver/openapi/python_examples.py:
```diff
@@ -1560,20 +1696,20 @@ def add_alert_words(client: Client) -> None:
 
 @openapi_test_function("/users/me/alert_words:delete")
 def remove_alert_words(client: Client) -> None:
+    words = client.get_alert_words()["alert_words"]
```
Yeah, I have added that now.

Vector73 commented on zerver/openapi/python_examples.py:
```diff
@@ -1605,7 +1741,7 @@ def update_user_group(client: Client, user_group_id: int) -> None:
     result = client.update_user_group(request)
     # {code_example|end}
-    assert result["result"] == "success"
+    validate_response_result(result)
```
No, I don't think so. Added that too.

Vector73 commented on zerver/openapi/python_examples.py:
```diff
@@ -100,7 +111,7 @@ def add_subscriptions(client: Client) -> None:
     )
     # {code_example|end}
-    assert result["result"] == "success"
+    validate_response_result(result)
```
No, I have added that now.

Vector73 commented:
I have addressed the recent feedback by @laurynmm and split the changes into multiple commits.

Vector73 commented on zerver/openapi/python_examples.py:
```diff
@@ -1109,7 +1109,7 @@ def get_attachments(client: Client) -> int:
     # {code_example|start}
-    # Delete the attachment with given attachment_id.
+    # Delete the attachment with given a ID.
 
```
This should be changed to "Delete the attachment given the attachment's ID."

Vector73 replied:
I have changed it to the comment mentioned above.

Vector73 commented on zerver/openapi/python_examples.py:
```diff
@@ -278,7 +278,7 @@ def create_user(client: Client) -> None:
 @openapi_test_function("/users/me/status:post")
 def update_status(client: Client) -> None:
     # {code_example|start}
-    # The request contains the new status and away boolean
+    # The request contains the new status and away boolean.
     request = {
```
I think this can be revised: `away -> "away"`?

Vector73 replied:
I changed this too. Do these changes look good?

Vector73 reviewed (commented):
Thanks @laurynmm for this clean-up. It looks great to me other than the comments below.

laurynmm commented:
@Vector73 - Thanks for making those updates. I think this looks good for now, so marking for integration review.

Follow-ups that remain are the error testing clean-ups:
- Remove `validate_against_openapi_schema` for the error tests since that's not currently testing anything. Probably replace those with an assertion/check of the error code (often defaults to `"BAD_REQUEST"`) that we expect to be returned for the error.
- Fixing the two error tests in `test-api` that are not returning errors in the test suite: `test_user_account_deactivated` and `test_realm_deactivated`.

timabbott commented:
Merged, thanks for all the work on this @Vector73 and @laurynmm! 

I really appreciate how much easier to maintain this now looks to be.
